ethereum/contracts/relayer/coreRelayer/CoreRelayerBase.sol

```diff
@@ -0,0 +1,264 @@
+// SPDX-License-Identifier: Apache 2
+
+pragma solidity ^0.8.19;
+
+import {IWormhole} from "../../interfaces/IWormhole.sol";
+import {IRelayProvider} from "../../interfaces/relayer/IRelayProvider.sol";
+import {toWormholeFormat, min, pay} from "./Utils.sol";
+import {
+  NoDeliveryInProgress,
+  ReentrantDelivery,
+  ForwardRequestFromWrongAddress,
+  RelayProviderDoesNotSupportTargetChain,
+  RelayProviderQuotedBogusAssetPrice,
+  Send,
+  DeliveryInstruction,
+  ExecutionParameters,
+  IWormholeRelayerBase
+} from "../../interfaces/relayer/IWormholeRelayer.sol";
+import {
+  ForwardInstruction,
+  DeliveryTmpState,
+  getDeliveryTmpState,
+  getRegisteredCoreRelayersState
+} from "./CoreRelayerStorage.sol";
+
+abstract contract CoreRelayerBase is IWormholeRelayerBase {
+  //TODO AMO: see https://book.wormhole.com/wormhole/3_coreLayerContracts.html#consistency-levels
+  //  15 is valid choice perhaps, but it feels messy... (seems like it should be 201?)
+  //  Also, these values should definitely not be defined here but should be provided by IWormhole!
+  uint8 internal constant CONSISTENCY_LEVEL_FINALIZED = 15;
+  uint8 internal constant CONSISTENCY_LEVEL_INSTANT = 200;
+
+  IWormhole private immutable wormhole_;
+  uint16 private immutable chainId_;
+
+  constructor(address _wormhole) {
+    wormhole_ = IWormhole(_wormhole);
+    chainId_ = uint16(wormhole_.chainId());
+  }
+
+  function getRegisteredCoreRelayerContract(uint16 chainId) public view returns (bytes32) {
+    return getRegisteredCoreRelayersState().registeredCoreRelayers[chainId];
+  }
+
+  //Our get functions require view instead of pure (despite not actually reading storage) because
+  //  they can't be evaluated at compile time. (https://ethereum.stackexchange.com/a/120630/103366)
+  
+  function getWormhole() internal view returns (IWormhole) {
+    return wormhole_;
+  }
+
+  function getChainId() internal view returns (uint16) {
+    return chainId_;
+  }
+
+  function publishAndPay(
+    uint256 wormholeMessageFee,
+    uint256 maxTransactionFee,
+    uint256 receiverValue,
+    bytes memory message,
+    uint8 consistencyLevel,
+    IRelayProvider relayProvider
+  ) internal returns (uint64 sequence) { 
+    sequence =
+      getWormhole().publishMessage{value: wormholeMessageFee}(0, message, consistencyLevel);
+
+    emit SendEvent(sequence, maxTransactionFee, receiverValue);
+
+    uint256 amount;
+    unchecked {amount = maxTransactionFee + receiverValue;}
+    //TODO AMO: what if pay fails? (i.e. returns false)
```

Comment:
Good point that if this fails, the value is stored in the contract. Is there an alternative you'd suggest?

Reply:
I'm happy to supply a backup address that is guaranteed to be payable :zany_face:

More seriously, I'm not sure what a good solution for this stupid edge case looks like.
We could have a mapping that keeps track of unredeemed funds per relay provider and provide a fallback mechanism to claim those funds, but the juice hardly seems worth the squeeze.

Probably the better solution is just to revert and have the relay provider fix their shit (since it's in their interest in the first place)

Reply:
Just trapping the funds seems like the worst of all worlds though. So I'd say: revert with a `RewardAddressNotPayable(address)` error

Reply:
Yeah, agreed, I had seen this before and was not sure about how to handle it. Reverting seemed excessive to me but at the same time, the reward address should be maintained extremely well, especially if it contains any code.